pySDC/projects/error_estimation/piline_hotrod_adaptivity.py

```diff
@@ -0,0 +1,169 @@
+    """
+
+    # initialize level parameters
+    level_params = dict()
+    level_params['dt'] = 3e-2
+    level_params['e_tol'] = 1e-8
+
+    # initialize sweeper parameters
+    sweeper_params = dict()
+    sweeper_params['collocation_class'] = CollGaussRadau_Right
+    sweeper_params['num_nodes'] = 3
+    sweeper_params['QI'] = 'IE'  # For the IMEX sweeper, the LU-trick can be activated for the implicit part
+    sweeper_params['QE'] = 'PIC'
+
+    problem_params = {
+        'Vs': 100.,
+        'Rs': 1.,
+        'C1': 1.,
+        'Rpi': 0.2,
+        'C2': 1.,
+        'Lpi': 1.,
+        'Rl': 5.,
+    }
+
+    # initialize step parameters
+    step_params = dict()
+    step_params['maxiter'] = 4
+
+    # initialize controller parameters
+    controller_params = dict()
+    controller_params['logger_level'] = 30
+    controller_params['hook_class'] = log_data
+    controller_params['use_HotRod'] = True
+    controller_params['use_adaptivity'] = use_adaptivity
+
+    # fill description dictionary for easy step instantiation
+    description = dict()
+    description['problem_class'] = piline  # pass problem class
+    description['problem_params'] = problem_params  # pass problem parameters
+    description['sweeper_class'] = imex_1st_order  # pass sweeper
+    description['sweeper_params'] = sweeper_params  # pass sweeper parameters
+    description['level_params'] = level_params  # pass level parameters
+    description['step_params'] = step_params
+
+    # set time parameters
+    t0 = 0.0
+    Tend = 2e1
+
+    # instantiate controller
+    controller_class = controller_nonMPI
+    controller = controller_class(num_procs=1, controller_params=controller_params,
+                                  description=description)
+
+    # get initial values on finest level
+    P = controller.MS[0].levels[0].prob
+    uinit = P.u_exact(t0)
+
+    # call main function to get things done...
+    uend, stats = controller.run(u0=uinit, t0=t0, Tend=Tend)
+    return stats
+
+
+def plot(stats, use_adaptivity):
+    setup_mpl()
+
+    # convert filtered statistics to list of iterations count, sorted by process
+    v1 = np.array(sort_stats(filter_stats(stats, type='v1'), sortby='time'))[:, 1]
+    v2 = np.array(sort_stats(filter_stats(stats, type='v2'), sortby='time'))[:, 1]
+    p3 = np.array(sort_stats(filter_stats(stats, type='p3'), sortby='time'))[:, 1]
+    t = np.array(sort_stats(filter_stats(stats, type='p3'), sortby='time'))[:, 0]
+    dt = np.array(sort_stats(filter_stats(stats, type='dt'), sortby='time'))[:, 1]
+    e_em = np.array(sort_stats(filter_stats(stats, type='e_embedded'), sortby='time'))[:, 1]
+    e_ex = np.array(sort_stats(filter_stats(stats, type='e_extrapolated'), sortby='time'))[:, 1]
+    ready = np.logical_and(e_ex != np.array(None), e_em != np.array(None))
+
+    assert np.allclose([v1[-1], v2[-2], p3[-1]], [83.88431516810506, 80.62596592922169, 16.134334413301595], rtol=1),\
```

Comment:
I don't love this, but this is the way to do it, right?

Reply:
Yes..